subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyConstraintsIntegrationTest.groovy

```diff
@@ -140,6 +140,42 @@ class DependencyConstraintsIntegrationTest extends AbstractIntegrationSpec {
         }
     }
 
+    void "dependency constraint does not preserve hard dependency for evicted dependency"() {
```

Comment:
It's very likely we have the exact same problem with optional dependencies.

Reply:
:o: I'd be tempted to add a comment on the signature of this test to highlight the fact it demonstrates the existing behavior and not the intended one. I have seen the comment in the expectations, but having it at the test declaration too makes it more readable as we usually don't dig into the details of a test.

Reply:
Thanks for the feedback. I updated the test to make it pass, added `@ToBeImplemented`, linked with `@Issue` and added a comment.